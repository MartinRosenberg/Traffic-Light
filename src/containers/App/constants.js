export const lights = Object.freeze(['red', 'yellow', 'green'])
