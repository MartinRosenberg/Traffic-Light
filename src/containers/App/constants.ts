export const lights: readonly string[] = ["red", "yellow", "green"]
