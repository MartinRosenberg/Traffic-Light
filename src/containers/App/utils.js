import { lights } from './constants'

export const fetchLight = () => {
  return new Promise((resolve, reject) => {
    const light = lights[Math.floor(Math.random() * 3)]
    light
      ? resolve(light)
      : reject(Error("Impossible condition"))
  })
}
