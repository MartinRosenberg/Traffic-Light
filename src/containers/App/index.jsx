import React, { Component } from 'react'

import { Light } from '../../components/Light'
import { Housing } from '../../components/Housing'
import { lights } from './constants'
import { fetchLight } from './utils'

export class App extends Component {
  state = {
    currentLight: null
  }

  handleLightChange = () =>
    fetchLight().then(result => {
      this.setState({ currentLight: result })
    }).catch(error => {
      console.error(error)
    })

  render () {
    return (
      <Housing
        lights={lights.map(light =>
          <Light
            color={light}
            key={light}
            powered={light === this.state.currentLight}
          />
        )}
        onClick={this.handleLightChange}
      />
    )
  }
}
