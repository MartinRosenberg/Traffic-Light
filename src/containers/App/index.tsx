import React, { FC, useState } from "react"

import { Light } from "../../components/Light"
import { Housing } from "../../components/Housing"
import { lights } from "./constants"
import { fetchLight } from "./utils"

type AppState = {
	currentLight: string | null
}

export const App: FC = () => {
	const [state, setState] = useState({ currentLight: null } as AppState)

	const handleLightChange = () =>
		fetchLight()
			.then((result) => {
				setState({ currentLight: result })
			})
			.catch((error) => {
				console.error(error)
			})

	return (
		<Housing
			lights={lights.map((light) => (
				<Light
					color={light}
					key={light}
					powered={light === state.currentLight}
				/>
			))}
			onClick={handleLightChange}
		/>
	)
}
