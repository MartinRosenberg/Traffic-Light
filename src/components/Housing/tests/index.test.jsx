import { shallow } from 'enzyme'
import React from 'react'

import { lights } from '../../../containers/App/constants'
import { Housing } from '..'

describe('Housing component', () => {
  test('should render correctly', () => {
    const onClick = jest.fn()
    const wrapper = shallow(<Housing lights={lights} onClick={onClick}/>)
    expect(wrapper).toMatchSnapshot()
  })
})
