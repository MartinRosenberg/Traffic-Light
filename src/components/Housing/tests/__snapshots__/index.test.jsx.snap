// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Housing component should render correctly 1`] = `
<div
  className="housing"
  onClick={[MockFunction]}
>
  red
  yellow
  green
</div>
`;
